File: 0016-3sum-closest/0016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
const threeSumClosest = (nums, target) => {
  nums.sort((a,b) => a - b);
  let closest_sum,
      sum,
      min_diff = Infinity;
  for (let a = 0; a < nums.length - 2; ++a) {
    let b = a + 1,
        c = nums.length - 1;
    while (b < c) {
      sum = nums[a] + nums[b] + nums[c];
      if (sum === target) return sum;
      sum < target ? ++b : --c;
      if (min_diff > Math.abs(target - sum)) {
        min_diff = Math.abs(target - sum);
        closest_sum = sum;
      }
    }
  }
  return closest_sum;
};

File: 0016-3sum-closest/0016-3sum-closest.go
import (
  "slices"
  "math"
)

func threeSumClosest(nums []int, target int) int {
  slices.Sort(nums)
  minDiff := math.MaxInt
  closestSum := 0
  for a, n := 0, len(nums); a < n - 2; a++ {
    b, c := a + 1, n - 1
    for b < c {
      sum := nums[a] + nums[b] + nums[c]
      if sum == target {
        return sum
      }
      diff := int(math.Abs(float64(target - sum)))
      if sum < target {
        b++
      } else {
        c--
      }
      if minDiff > diff {
        minDiff = diff
        closestSum = sum
      }
    }
  }
  return closestSum
}